### PR TITLE
feat: add XMLHttpRequest instrumentation with trace-context propagation

### PR DESCRIPTION
**Pull request by parker-edwards**
## What

Adds `XMLHttpRequest` instrumentation (`@dash0/xhr`) — HTTP client spans and trace-context propagation for XHR, matching the existing fetch instrumentation. This replaces #93 (closed), split per feature at the reviewing team's request — this is part 2 of 3 and independent of the startView PR (#96); the interaction-instrumentation PR (#98) is stacked on top of this one.

## Why

axios's default browser adapter uses `XMLHttpRequest`, so today axios/XHR-based applications get no HTTP spans and no `traceparent` header — their frontend telemetry never joins backend traces.

## How

- New `@dash0/xhr` instrumentation name, enabled by default alongside `@dash0/fetch`.
- First commit extracts the propagator/error helpers from `fetch.ts` into `http/utils.ts`, so fetch and XHR share one propagation code path (`traceparent`, X-Ray, `propagators` config, CORS URL allow-lists) — no behavior change for fetch.
- XHR spans cover the full lifecycle: success, error, timeout, and abort all end the span, and per-request event listeners are removed on completion to avoid leaks.
- Documented in `INSTALL.md` and the README feature list.
- `axios` is added as a devDependency only, for the e2e spec.

## Known limitation (candidate follow-up)

Environments that polyfill `fetch` on top of XHR will produce two spans for one logical request. A polyfill guard is a natural follow-up if you want it; I kept it out of scope here.

## Testing

- `pnpm run ci` (lint, prettier, unit tests, build): green — 275 unit tests across 17 files, including 16 new XHR unit tests.
- New e2e spec `09-xhr-instrumentation` (raw XHR, cross-origin propagation, X-Amzn-Trace-Id delivery, axios): green locally via `pnpm run test:e2e:local`.

## Notes for reviewers

This is my own initiative rather than a scheduled roadmap item — happy to adjust naming, semantics, or scope to fit your conventions.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by thiemok — COMMENTED** [review]
Extensive review of the XHR instrumentation, covering general correctness, host-page safety, consistency with the fetch instrumentation, docs, and a comparison against OTel JS, Grafana Faro, and Instana weasel. Inline comments follow the [conventional comments](https://conventionalcomments.org/) format.

**praise:** The overall architecture is strong and in several dimensions ahead of the OTel XHR reference: creating the span at `send()` (not `open()`) avoids OTel's span-leak class for open-without-send; a single `loadend` completion point with a re-entry guard is simpler and more complete than OTel's four separate end listeners; resource-timing correlation drives the actual span duration; URL hygiene (userinfo redaction + scrubber + `url.*` decomposition) beats all three reference SDKs; abort is a first-class non-error; own-beacon exclusion is built in; prototype patching preserves `instanceof`/subclassing/Zone.js interop; and symbol-keyed per-instance state is invisible to page code. Unit coverage is a superset of what fetch has today, and the `utils.ts` extraction is byte-identical/behavior-preserving (verified by diff). The two blocking items below are both cheap fixes in the "must never break the host page" bucket — which matters extra since `@dash0/xhr` is enabled by default for upgrading users.

Items that don't fit an inline anchor:

**suggestion (non-blocking):** E2E coverage gaps vs the fetch suite: there is no XHR e2e case for error spans, abort, or request-body passthrough (fetch has "must send body from init/Request"; an XHR equivalent would verify `send(body)` arrives unchanged). Sync XHR is only tested for the success path.

**suggestion (non-blocking, docs):** INSTALL.md's `headersToCapture` entry says headers are "matched against HTTP request headers", but both instrumentations also capture **response** headers — and for XHR, cross-origin response headers are only readable when the server sets `Access-Control-Expose-Headers` (the code even logs a CORS debug message). Suggest "request and response headers" plus the expose-headers caveat. The error/timeout/abort span semantics are also undocumented for both instrumentations — optional, but this PR would be a natural place to add a short behavior reference.

**thought (non-blocking, follow-ups):** Two ideas from the peer comparison worth tickets rather than this PR: (1) Instana caps monitored XHRs (32/10s, 256/10min) as a backstop against pathological polling pages — `src/utils/rate-limit.ts` exists but isn't applied to HTTP spans (fetch or XHR); (2) OTel emits `server.address`/`server.port` per HTTP-client semconv where we only emit `url.domain` — if ever added, add it in shared code so fetch gets it too.

File: src/instrumentations/http/xhr.ts [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  wrap(proto, "open", wrapOpen);
+  wrap(proto, "setRequestHeader", wrapSetRequestHeader);
+  wrap(proto, "send", wrapSend);
+}
+
+function wrapOpen(original: XMLHttpRequest["open"]): XMLHttpRequest["open"] {
+  return function (this: InstrumentedXhr, method: string, url: string | URL, ...rest: unknown[]) {
+    const stringUrl = String(url);
+    const originalMethod = method ?? "GET";
+    const isWellKnownMethodMatchingLeniently = isWellKnownHttpMethod(originalMethod.toUpperCase());
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **issue (blocking):** SDK code can synchronously throw out of the page's `xhr.open()`/`xhr.send()` — there is no enclosing try/catch in `wrapOpen`/`wrapSend`, and all SDK work runs *before* the original method is called. Two verified throw paths:

1. **Config typos become a page-wide XHR outage.** `isUrlIgnored` → `matchesAny` (`ignore-rules.ts`), `determinePropagatorTypes` (`utils.ts`), and `vars.headersToCapture.some((rxp) => rxp.test(name))` all call `.test()` unguarded. A plain-JS consumer passing strings instead of RegExp (e.g. `ignoreUrls: ["/health"]`) makes **every** `xhr.open()` on the page throw. Fetch has the same exposure, but there it surfaces as a rejected promise, which pages routinely handle — a sync throw from `open()` is the worst case for host-page safety.
2. **Non-string `method` throws where native XHR doesn't.** `originalMethod.toUpperCase()` throws `TypeError` on `xhr.open(123, "/x")`, which native `open()` accepts via ByteString conversion (method `"123"`). `url` is already defended with `String(url)`; `method` should be too (`String(method ?? "GET")` — note `String(url)` also evaluates a custom `toString`/`Symbol.toPrimitive` twice, SDK + native, so side effects run twice; worth a look while here).

Suggest wrapping the SDK-side blocks of both wrappers in try/catch that falls through to `original.apply(this, args)` — one belt-and-braces guard makes every internal error non-fatal (standard RUM-SDK posture; OTel and Instana both guard here), and doing it in `wrap()`/shared code would cover fetch as well. Related: the prototype assignment in `instrumentXhr()` itself is unguarded — a page that locked `XMLHttpRequest.prototype` via `Object.defineProperty` makes `init()` throw and abort the remaining instrumentations (OTel try/catches + unwraps for exactly this, with a warning naming that scenario).

File: src/instrumentations/http/xhr_test.ts [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    xhr.respond(200);
+
+    const sendSpanMock = sendSpan as unknown as ReturnType<typeof vi.fn>;
+    await vi.waitFor(() => expect(sendSpanMock).toHaveBeenCalledTimes(1));
+    const span = sendSpanMock.mock.calls[0]![0] as Span;
+    expect(span.name).toBe("HTTP GET");
+    expect(span.attributes).toContainEqual({ key: "http.request.method", value: { stringValue: "GET" } });
+  });
+
+  it("is safe to call instrumentXhr() twice (double-instrumentation guard)", async () => {
```
Comment (thiemok) [inline src/instrumentations/http/xhr_test.ts:429]: **issue (blocking):** The double-instrumentation guard this test claims does not actually exist, and the test passes for the wrong reason. `wrap()` (`src/utils/wrap.ts:27-33`) marks the **original** function as instrumented, then replaces it with an **unmarked** wrapper — so a second `wrap()` call sees no marker and wraps again. Verified empirically: after two `instrumentXhr()` calls, one `send()` produces 8 listener registrations and **2 `traceparent` injections**. Per the XHR spec, repeated `setRequestHeader()` calls with the same name *combine* values, so the customer's backend receives `traceparent: 00-a…-01, 00-b…-01` — an invalid header that breaks their server-side trace propagation (and possibly strict validation/WAFs).

The `Symbol.for("INSTRUMENTED_BY_DASH0")` global-registry key shows cross-copy dedupe was the intent (script-tag + npm bundle, micro-frontends) — `init()`'s `hasBeenInitialised` only protects a single module instance. This test passes only because both wrappers share the same module-level `XHR_STATE` symbol, so `state.completed` collapses everything to one `sendSpan` — it doesn't detect the doubled header and can't model the two-copy case (separate symbols → full duplication).

The `wrap.ts` flaw is pre-existing (affects fetch too), but this PR triples the wrapped surface and adds a test enshrining a guarantee that doesn't hold. Fix is one line in `wrap.ts` — mark the wrapper after assignment (keep the original marked too) — plus strengthening this test to count listener additions or injected header values.

File: src/instrumentations/http/xhr.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+    // A new open() call on a reused XHR instance resets state -- any prior span for this instance
+    // has already been sent (or will never be, if the previous request never completed) and this
+    // open() is treated as the start of a brand-new request with its own span.
+    this[XHR_STATE] = {
+      method: normalizedMethod,
+      originalMethod,
+      isWellKnownMethod: isWellKnownHttpMethod(originalMethod),
+      url: stringUrl,
+      ignored: isUrlIgnored(stringUrl),
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **issue:** `isUrlIgnored` receives the raw, possibly-relative URL, while fetch matches against the resolved absolute URL (`Request` resolves it before `fetch.ts:52`). Consequences:

- An origin-anchored ignore regex (e.g. `/^https:\/\/myapp\.com\/health/`) suppresses `fetch("/health")` but **not** the identical `xhr.open("GET", "/health")`; conversely `/^\/internal\//` matches XHR but can never match fetch. INSTALL.md documents `ignoreUrls` as applying uniformly to "XMLHttpRequest, fetch and resource URLs".
- `isTransportRequest`'s `startsWith(endpoint.url)` self-exclusion can never match a relative XHR URL (edge case, since the SDK transport itself uses fetch).

Suggest resolving first — `isUrlIgnored(parseUrl(stringUrl).href)` — as this file already does for resource matching in `wrapSend`. Note the e2e ignore test doesn't catch this because its regex is a path substring that matches both forms.

**nitpick (non-blocking):** for parity, fetch also emits a `debug("not creating span ... ignored")` log on this path; XHR is silent.

File: src/instrumentations/http/xhr.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  return function (this: InstrumentedXhr, method: string, url: string | URL, ...rest: unknown[]) {
+    const stringUrl = String(url);
+    const originalMethod = method ?? "GET";
+    const isWellKnownMethodMatchingLeniently = isWellKnownHttpMethod(originalMethod.toUpperCase());
+    const normalizedMethod = isWellKnownMethodMatchingLeniently ? originalMethod.toUpperCase() : HTTP_METHOD_OTHER;
+
+    // A new open() call on a reused XHR instance resets state -- any prior span for this instance
+    // has already been sent (or will never be, if the previous request never completed) and this
+    // open() is treated as the start of a brand-new request with its own span.
+    this[XHR_STATE] = {
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **issue:** The comment covers the dropped span, but reopen-while-in-flight leaks more than that. Per spec, `open()` during an in-flight request terminates the fetch **without firing `abort`/`loadend`** — so the previous state's `onLoadEnd` never runs: its 4 listeners stay attached to the instance, its span never ends, and its `PerformanceObserver` keeps buffering until the 10-minute disposer. The common cancel-by-reopen polling pattern on a reused XHR accumulates this every cycle. Worse, the stale `loadend` listener from the old `send()` fires on the *next* request's completion and `onLoadEnd(this, oldState)` ends the old span using the new request's status.

Suggest: when `wrapOpen` finds a previous non-`completed` state, clean it up — remove its listeners, `performanceObserver?.cancel()`, and arguably end its span as cancelled (`endSpanOnAbort`), matching how fetch reports cancelled requests instead of silently dropping them.

Relatedly, `wrapSend` doesn't guard against `send()` being called twice: the second call creates a second span and 4 more listeners *before* the native `InvalidStateError` fires, overwriting `state.span`/`state.listeners`/`state.performanceObserver` — the response then gets attributed to the second span (wrong start time) and the first leaks. A cheap guard: if `state.span && !state.completed`, just `return original.call(this, body)`.

File: src/instrumentations/http/xhr.ts [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+          state.failureKind = "abort";
+        },
+      ],
+      ["loadend", () => onLoadEnd(this, state)],
+    ];
+    for (const [eventType, listener] of state.listeners) {
+      addEventListener(this, eventType, listener);
+    }
+
+    return original.call(this, body);
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **issue:** For synchronous XHR (`open(..., false)`), a network error or timeout makes `send()` **throw** instead of firing events — per the spec's sync error handling, `loadend` does not fire. The span is then never ended or sent, the 4 listeners stay attached, and the observer lives for 10 minutes. The e2e suite covers sync *success* only.

Fetch handles the analogous rejection path (`endSpanOnError` on the promise rejection); suggest a try/catch here that cancels the observer, removes the listeners, ends the span as an error, and rethrows the original exception unchanged.

File: src/instrumentations/http/xhr.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+function wrapSetRequestHeader(original: XMLHttpRequest["setRequestHeader"]): XMLHttpRequest["setRequestHeader"] {
+  return function (this: InstrumentedXhr, name: string, value: string) {
+    original.call(this, name, value);
+
+    const state = this[XHR_STATE];
+    if (!state || state.ignored) return;
+
+    (state.capturedRequestHeaders ??= {})[name] = value;
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **issue (non-blocking):** Every request header name/value is retained on the instance unconditionally — including `Authorization`, cookies, CSRF tokens — even when `vars.headersToCapture` is empty (the common case). It's never transmitted unless matched, but it's needless credential retention on a page-reachable object until the next `open()`/GC; fetch filters at capture time and retains nothing. Suggest gating on `vars.headersToCapture.length > 0` before storing (the read side in `wrapSend` already checks it).

Two semantic edge cases while here: (1) native XHR *combines* repeated `setRequestHeader` calls for the same name (case-insensitively) into `"a, b"`, but this map overwrites for identical casing and stores separate entries for differing casings — the latter then produce duplicate attribute keys after `httpRequestHeaderKey` lowercases them, since `addAttribute` pushes without dedup; (2) matching runs against the raw header name, so a case-sensitive regex like `/^X-Test/` captures on XHR but not on fetch (where `Headers` iteration yields lowercased names).

File: src/instrumentations/http/xhr.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    addAttribute(span.attributes, HTTP_REQUEST_METHOD, state.method);
+    if (!state.isWellKnownMethod) {
+      addAttribute(span.attributes, HTTP_REQUEST_METHOD_ORIGINAL, state.originalMethod);
+    }
+    state.span = span;
+
+    if (state.propagatorTypes.length > 0) {
+      try {
+        addTraceContextHttpHeaders(
+          (name, value) => this.setRequestHeader(name, value),
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **suggestion:** Two things about injecting via the **wrapped** `this.setRequestHeader`:

1. The injected `traceparent`/`X-Amzn-Trace-Id` land in `state.capturedRequestHeaders` and become span attributes under a broad `headersToCapture` pattern — fetch's `copyOfInit` path doesn't capture SDK-injected headers. Injecting via the original function (Instana does this) or injecting before wiring capture would restore parity.
2. This is also the natural place to defuse **fetch-polyfill double instrumentation**: `fetch.ts` deliberately matches `initiatorType === "xmlhttprequest"` "to support polyfills" — on a page where fetch is polyfilled over XHR, one logical request now produces **two spans** and a double header injection (repeated `setRequestHeader` *combines* values → invalid combined `traceparent` at the customer's backend). Instana handles exactly this case: its `setRequestHeader` wrap detects an already-present correlation header and marks the XHR ignored. Since we already wrap `setRequestHeader` and track state, skipping span creation (or at least injection) in `wrapSend` when a traceparent/X-Ray header was already set on the instance is nearly free.

File: src/instrumentations/http/utils.ts [hunk 8]
```diff
@@ -47,3 +60,64 @@ export function addResourceSize(span: InProgressSpan, resource: PerformanceResou
     addAttribute(span.attributes, HTTP_RESPONSE_BODY_SIZE, encodedLength);
   }
 }
+
+export function endSpanOnError(span: InProgressSpan, error: Exception) {
```
Comment (thiemok) [inline src/instrumentations/http/utils.ts:70]: **suggestion (non-blocking):** The error-telemetry shape now diverges between the two instrumentations for the same failure class. Fetch (via this helper) records the *real* exception — `exception.type` = e.g. `TypeError`, real message/stack — but sets **no** `error.type` attribute; XHR sets `error.type` = `error`/`timeout` with a synthetic exception. Also: fetch's resolved `status === 0` path sets `error.type = response.type` (e.g. `opaque`) *and* records `http.response.status_code: "0"`, while XHR's `status === 0` takes the failure branch with no status attribute; and an XHR timeout is an ERROR span while `AbortSignal.timeout()` on fetch surfaces as a benign cancellation (inherent to the APIs, but worth documenting).

Net effect: querying failed HTTP spans by `error.type` returns only XHR failures. Suggest aligning — the cheapest fix is adding an `error.type` here in `endSpanOnError` (derived from the exception name) so fetch matches XHR's queryability.

File: src/instrumentations/http/xhr.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    addAttribute(span.attributes, ERROR_TYPE, failureKind);
+    sendSpan(endSpan(span, { code: SPAN_STATUS_ERROR, message: `XMLHttpRequest failed: ${state.url}` }, undefined));
+    return;
+  }
+
+  setSpanStatus(span, status >= 200 && status < 400 ? SPAN_STATUS_UNSET : SPAN_STATUS_ERROR);
+  addAttribute(span.attributes, HTTP_RESPONSE_STATUS_CODE, String(status));
+  tryCaptureResponseHeaders(xhr, span);
+
+  if (!performanceObserver) {
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **nitpick (non-blocking):** Dead branch — `state.performanceObserver` is unconditionally set in `wrapSend` before any listener can fire, so `performanceObserver` can't be undefined here.

File: src/instrumentations/http/xhr.ts [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+        const name = line.substring(0, separatorIndex).trim();
+        const value = line.substring(separatorIndex + 1).trim();
+
+        if (vars.headersToCapture.some((rxp) => rxp.test(name))) {
+          addAttribute(span.attributes, httpResponseHeaderKey(name), value);
+        }
+      });
+  } catch (_e) {
+    debug("unable to capture http response headers due to CORS policy");
```
Comment (thiemok) [inline src/instrumentations/http/xhr.ts]: **nitpick (non-blocking):** Misleading message — `getAllResponseHeaders()` doesn't throw for CORS; it silently filters non-exposed headers (cross-origin you just get fewer lines unless `Access-Control-Expose-Headers` is set). The actual error `_e` is also discarded; suggest logging it and a neutral message. (Copied from fetch's equally misleading message, so fixing both would be consistent.)

File: src/instrumentations/http/xhr_test.ts [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+    const sendSpanMock = sendSpan as unknown as ReturnType<typeof vi.fn>;
+    await vi.waitFor(() => expect(sendSpanMock).toHaveBeenCalledTimes(1));
+    const span = sendSpanMock.mock.calls[0]![0] as Span;
+    expect(span.attributes).toContainEqual({
+      key: "http.request.header.x-test-header",
+      value: { stringValue: "hello" },
+    });
+  });
+
+  it("normalizes well-known methods to uppercase and records HTTP_METHOD_OTHER for unknown methods", async () => {
```
Comment (thiemok) [inline src/instrumentations/http/xhr_test.ts:400]: **todo (non-blocking):** The title promises `HTTP_METHOD_OTHER` coverage, but the test only opens with lowercase `"get"` — the `_OTHER` normalization and the `http.request.method_original` attribute are untested in both unit and e2e suites. Add a case with a non-well-known method (e.g. `"FROBNICATE"`) asserting `http.request.method: _OTHER` + `http.request.method_original`.

File: test/e2e/spec/09-xhr-instrumentation/09-xhr-instrumentation.test.ts [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    await expect(browser).toHaveTitle("xhr instrumentation test");
+
+    const ignoredBtn = await $("button=XHR With Ignored URL");
+    await ignoredBtn.click();
+
+    const okBtn = await $("button=Same Origin XHR");
+    await okBtn.click();
+
+    await retry(async () => {
+      await expectSpanCount(1);
```
Comment (thiemok) [inline test/e2e/spec/09-xhr-instrumentation/09-xhr-instrumentation.test.ts]: **issue (non-blocking):** This assertion can't reliably detect a broken ignore: `expectSpanCount(1)` counts OTLP `/v1/traces` **requests**, not spans, and the batcher merges everything sent within a 1s window into one request. The two clicks happen back-to-back, so if ignoring were broken, both spans would very likely land in a single batch and the test would still pass. Assert instead that no span with `url.path` matching `you-cant-see-this` exists (keeping the positive match).

File: test/e2e/spec/09-xhr-instrumentation/09-xhr-instrumentation.test.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    await loadPage("/e2e/spec/09-xhr-instrumentation/page.html");
+    await expect(browser).toHaveTitle("xhr instrumentation test");
+
+    await browser.execute(async () => {
+      await fetch("http://localhost.lambdatest.com:8012/ajax-requests", { method: "DELETE" }).catch(() => {});
+    });
+
+    const btn = await $("button=Cross Origin XHR");
+    await btn.click();
+    await sleep(500);
```
Comment (thiemok) [inline test/e2e/spec/09-xhr-instrumentation/09-xhr-instrumentation.test.ts]: **nitpick (non-blocking):** Bare `sleep(500)` instead of the suite's `retry` pattern makes this timing-sensitive, and the test asserts only that the propagation header reached the server — never that a span was produced for the cross-origin request. Suggest `retry` + an `expectSpanMatching` for the cross-origin URL.

File: src/instrumentations/http/xhr.ts (hunk 1 above)
Comment (thiemok) [reply src/instrumentations/http/xhr.ts]: Addressed in 91c235a:

- **Sync throws out of `open()`/`send()`**: `wrapOpen`/`wrapSetRequestHeader`/`wrapSend` are now thin shells — the SDK work lives in `onOpen`/`onSetRequestHeader`/`onSend` and is guarded by try/catch that always falls through to the original method (invoked exactly once, outside the guard). `wrapSend`'s catch also does best-effort cleanup (cancel observer, remove listeners, drop the span) so a stray `loadend` can't emit a half-initialized span.
- **Non-string method / double URL coercion**: method is now coerced via `String(method ?? "GET")` for SDK bookkeeping, and the pre-coerced URL string is handed to the native `open()` so a side-effectful custom `toString` runs once instead of twice (the method arg is passed through untouched).
- **Locked prototype**: the guard went into `wrap()` itself rather than `instrumentXhr()`, so a locked property logs a warning and `init()` proceeds with the remaining instrumentations — this covers fetch and future `wrap()` users too. A generic *call-path* guard in `wrap()` isn't safe (it can't know whether the original was already invoked, e.g. `wrapSetRequestHeader` calls it first), so those stay inline per wrapper.
- **Fetch**: got the same fall-through — instrumentation-prep failures now degrade to an uninstrumented `original(...)` call instead of a rejected promise.
- Bonus: the new `wrap_test.ts` surfaced a pre-existing bug — the double-instrumentation guard marked only the original function, so repeated `wrap()` calls double-wrapped. The replacement is now marked as well.

Covered by new unit tests (string-instead-of-RegExp configs, non-string method, single `toString` evaluation, throwing header matchers, locked prototypes); full unit + local e2e suites pass.

File: src/instrumentations/http/xhr.ts (hunk 3 above)
Comment (thiemok) [reply src/instrumentations/http/xhr.ts]: Fixed in ba4aa74. `onOpen` now resolves the URL once via `parseUrl(stringUrl).href` and uses the absolute form for all SDK bookkeeping — the ignore check, `determinePropagatorTypes` (which had the same raw-vs-resolved inconsistency for cross-origin `match` regexes), `url.*` attributes, and the resource matcher (simplified to `name === state.url` since state now holds the resolved href). The native `open()` still receives the page's original URL untouched, and a `parseUrl` throw degrades to an uninstrumented request via the existing `wrapOpen` catch, mirroring fetch's behavior when `new Request()` throws.

Also added the debug log on the ignored path (in `send`, mirroring fetch's "not creating span" semantics) and unit tests pinning that an origin-anchored ignore regex suppresses a relative-URL XHR and that `url.full` is the resolved absolute URL.

File: src/instrumentations/http/xhr_test.ts (hunk 2 above)
Comment (thiemok) [reply src/instrumentations/http/xhr_test.ts:429]: Fixed across two commits:

- **91c235a** fixes the `wrap()` flaw: the wrapped replacement is now marked as instrumented alongside the original, so a second `wrap()` call is a no-op. Also adds `src/utils/wrap_test.ts` with a "does not wrap the same target twice" test (two `wrap()` calls → wrapper factory invoked exactly once).
- **752e338** strengthens this test so it can detect the doubled-injection failure mode directly: `FakeXMLHttpRequest.setRequestHeader` now combines repeated values per the XHR spec, and the double-instrumentation test enables a traceparent propagator and asserts the injected header matches a single well-formed value (a regression would produce a comma-joined header and fail the match). Verified by temporarily removing the `markAsInstrumented(wrapped)` lines — both this test and the wrap unit test fail.

File: src/instrumentations/http/xhr.ts (hunk 4 above)
Comment (thiemok) [reply src/instrumentations/http/xhr.ts]: Addressed in e766d74:

- **Reopen while in flight:** `onOpen` now detects a previous non-`completed` state with a span and cleans it up before resetting — removes its listeners, cancels its `PerformanceObserver`, and ends the span as cancelled via `endSpanOnAbort` (matching how fetch reports cancelled requests). The cleanup runs at the very top of `onOpen`, before URL coercion, so it also happens when a custom `toString` throws.
- **Double `send()`:** `wrapSend` now returns straight into the native call when `state.span && !state.completed`, so the second call (which throws `InvalidStateError` natively) no longer overwrites the in-flight request's span/listeners/observer. Retry after a failed instrumented send still works since `cleanupFailedSend` clears `state.span`.

Unit tests added for both: reopen-while-in-flight asserts the old span is sent as cancelled without a status code and the stale loadend listener doesn't re-end it with the next request's status; double-send asserts exactly one span and one set of listeners.

File: src/instrumentations/http/xhr.ts (hunk 5 above)
Comment (thiemok) [reply src/instrumentations/http/xhr.ts]: Fixed in c373b38. `wrapSend` now wraps the native `send()` call in a try/catch: on a synchronous throw it removes the four per-request listeners, cancels the performance observer, records the thrown exception on the span, sets `error.type` (`timeout` for a `TimeoutError` DOMException, `error` otherwise, matching the async path's classification), ends the span with error status, and rethrows the page's original exception unchanged.

The double-`send()` guard path intentionally keeps its bare `original.call` — the native `InvalidStateError` there must not tear down the still-in-flight request's listeners and span. The cleanup itself is guarded by `state.completed` and its own try/catch so it can neither double-finalize a span nor mask the page's exception.

Covered by three new unit tests: sync network error ends the span as an error, removes all listeners, and rethrows the exact same exception object; `TimeoutError` is classified as `timeout`; and the instance still instruments a fresh request after a sync failure.

File: src/instrumentations/http/xhr.ts (hunk 6 above)
Comment (thiemok) [reply src/instrumentations/http/xhr.ts]: Addressed in c1d5915. `setRequestHeader` now filters at capture time like the fetch instrumentation: nothing is retained when `headersToCapture` is empty, and only matching headers are ever stored — so credentials like `Authorization` no longer sit on the page-reachable instance.

Both edge cases are covered too: repeated `setRequestHeader()` calls now combine case-insensitively into a single `"a, b"` value under the lowercased key (no more duplicate attribute keys), and matching runs against the lowercased header name so regexes behave identically for XHR and fetch (documented in INSTALL.md).

One behavioral consequence: a throwing capture matcher now throws inside the wrapped `setRequestHeader` (where it's caught) instead of inside `send()` — so it no longer aborts span creation; the request is tracked normally and only the header capture is skipped.

File: src/instrumentations/http/xhr.ts (hunk 7 above)
Comment (thiemok) [reply src/instrumentations/http/xhr.ts]: Addressed in 53b1fa3:

1. Trace context is now injected via the un-wrapped `setRequestHeader` (captured at wrap time inside the `wrapSetRequestHeader` factory, so double instrumentation can't capture our own wrapper). SDK-injected headers no longer land in `capturedRequestHeaders` / span attributes, matching fetch.
2. `setRequestHeader` now marks the request as already traced when a `traceparent` / `X-Amzn-Trace-Id` header is set externally (case-insensitive; our own injection bypasses the wrapper, so it can't trip this). `send()` then skips both span creation and injection — a fetch polyfill over XHR yields one span (the fetch one, whose resource matcher already accepts `initiatorType === "xmlhttprequest"`) and one valid traceparent. `open()` resets the flag, so reused instances are traced normally again.

Known gap, kept deliberately (same as Instana): this only defuses the polyfill case when correlation headers were actually injected (a propagator matched the URL) — without a match, one logical request still produces two spans. A synchronous "inside instrumented fetch" context flag could close that, but we decided to leave it out of this PR.

File: src/instrumentations/http/utils.ts (hunk 8 above)
Comment (thiemok) [reply src/instrumentations/http/utils.ts:70]: Addressed in d5cc46553c28dc0e423c577d8340f06bd1bd2a4b: `endSpanOnError` now sets `error.type` derived from the exception name (falling back to `code`, then `"error"`), so failed fetch spans are queryable by `error.type` like XHR failures. The XHR loadend failure branch now reuses the helper (behavior-identical — its synthetic exception carries the failure kind as the name), and the divergences inherent to the APIs (fetch's opaque status-0 shape, `AbortSignal.timeout()` surfacing as a cancellation vs XHR's timeout ERROR span) are documented on `endSpanOnError`/`endSpanOnAbort`.

**Comment by thiemok** [comment]
Closing in favor of #99, which is the same branch and commits re-opened from an in-repo branch — fork PRs don't receive the LambdaTest secrets, so the required e2e checks could never pass here. Your commits and authorship are carried over unchanged, and the squash merge will credit you via Co-authored-by. Thanks again for the contribution, @parker-edwards!
